{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f76f367b-1ad4-4e3f-bbc8-dabf0de74e41",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Load libraries\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import mygene\n",
    "import os\n",
    "\n",
    "## Display all rows of pandas dataframes\n",
    "pd.set_option('display.max_rows', None)\n",
    "pd.set_option('display.max_colwidth', None)\n",
    "\n",
    "# Opt into the future behavior for silent downcasting\n",
    "pd.set_option('future.no_silent_downcasting', True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "815b76b5-c27a-4d1a-8f12-a82e982a83f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def unify_gene_ids_and_symbols(df):\n",
    "\n",
    "    ## Make a copy to use in function\n",
    "    dff = df.copy()\n",
    "    \n",
    "    ## Get all the unique gene_id and gene_symbols\n",
    "    df_ids = dff[\"gene_id\"].copy().drop_duplicates(inplace=False).dropna(inplace=False).to_list()\n",
    "    df_symbols = dff[\"gene_symbol\"].copy().drop_duplicates(inplace=False).dropna(inplace=False).to_list()\n",
    "\n",
    "    ## Query by gene_ids and gene_symbols\n",
    "    mg = mygene.MyGeneInfo()\n",
    "    results_id_query = mg.querymany(df_ids, scopes='symbol,alias,name,ensembl.gene,ensembl.transcript', fields='symbol,ensembl.gene', species='human', as_dataframe=True, df_index=False, returnall=False)\n",
    "    results_symbol_query = mg.querymany(df_symbols, scopes='symbol,alias,name,ensembl.gene,ensembl.transcript', fields='symbol,ensembl.gene', species='human', as_dataframe=True, df_index=False, returnall=False)\n",
    "    \n",
    "    ## Delete duplicates, only keep highest value\n",
    "    results_id_query = results_id_query.sort_values(by=['query', '_score'], ascending=[True, False]).drop_duplicates(subset='query', keep='first')\n",
    "    results_symbol_query = results_symbol_query.sort_values(by=['query', '_score'], ascending=[True, False]).drop_duplicates(subset='query', keep='first')\n",
    "    \n",
    "    # Ensure that the column 'ensembl' is of string type, which allows the use of string methods\n",
    "    results_id_query[\"ensembl\"] = results_id_query[\"ensembl\"].astype(str)\n",
    "    results_symbol_query[\"ensembl\"] = results_symbol_query[\"ensembl\"].astype(str)\n",
    "    \n",
    "    # Split the strings and handle NaNs by chaining the .str methods\n",
    "    results_id_query[\"first_ensembl\"] = results_id_query[\"ensembl\"].str.split(\":\", expand=True)[1].str.split(\"'\", expand=True)[1]\n",
    "    results_symbol_query[\"first_ensembl\"] = results_symbol_query[\"ensembl\"].str.split(\":\", expand=True)[1].str.split(\"'\", expand=True)[1]\n",
    "\n",
    "    # Standardize all NaN-like values to pd.NA\n",
    "    results_id_query = results_id_query.replace({np.nan: pd.NA, None: pd.NA, \"nan\": pd.NA})\n",
    "    results_symbol_query = results_symbol_query.replace({np.nan: pd.NA, None: pd.NA, \"nan\": pd.NA})\n",
    "    \n",
    "    # Fill NA values in 'ensembl.gene' column with values from 'first_ensembl' column\n",
    "    results_id_query['ensembl.gene'] = results_id_query['ensembl.gene'].copy().fillna(results_id_query['first_ensembl'], inplace=False)\n",
    "    results_symbol_query['ensembl.gene'] = results_symbol_query['ensembl.gene'].copy().fillna(results_symbol_query['first_ensembl'], inplace=False)\n",
    "\n",
    "    ## Only keep relevant columns\n",
    "    results_id_query = results_id_query[[\"query\", \"ensembl.gene\", \"symbol\"]].copy()\n",
    "    results_symbol_query = results_symbol_query[[\"query\", \"ensembl.gene\", \"symbol\"]].copy()\n",
    "\n",
    "    ## Rename columns\n",
    "    results_id_query.columns = [\"gene_id\", \"from_id_ensembl_id\", \"from_id_official_gene_symbol\"]\n",
    "    results_symbol_query.columns = [\"gene_symbol\", \"from_symbol_ensembl_id\", \"from_symbol_official_gene_symbol\"]\n",
    "\n",
    "    results_id_query.drop_duplicates(inplace=True)\n",
    "    results_symbol_query.drop_duplicates(inplace=True)\n",
    "\n",
    "    \n",
    "    ## Merge with original dataframe\n",
    "    dff = dff.merge(results_id_query, how=\"left\", on=\"gene_id\")\n",
    "    dff = dff.merge(results_symbol_query, how=\"left\", on=\"gene_symbol\")\n",
    "\n",
    "    ## Combine results from both approaches (symbol and id)\n",
    "    dff['ensembl_id'] = dff['from_id_ensembl_id'].copy().fillna(dff['from_symbol_ensembl_id'], inplace=False)\n",
    "    dff['official_symbol'] = dff['from_id_official_gene_symbol'].copy().fillna(dff['from_symbol_official_gene_symbol'], inplace=False)\n",
    "\n",
    "    ## Drop intermediate columns\n",
    "    dff.drop(columns=[\"from_id_ensembl_id\", \"from_id_official_gene_symbol\", \"from_symbol_ensembl_id\", \"from_symbol_official_gene_symbol\", \"gene_id\", \"gene_symbol\"], inplace=True)\n",
    "\n",
    "    ## Drop anything that is missing both ensembl_id and official_symbol. Also drop any duplicates\n",
    "    dff.dropna(subset=[\"ensembl_id\", \"official_symbol\"], how=\"all\", inplace=True)\n",
    "    \n",
    "    ## Drop Duplicates\n",
    "    dff.drop_duplicates(inplace=True)\n",
    "    dff.drop_duplicates(subset=[\"ensembl_id\", \"official_symbol\"], inplace=True)\n",
    "\n",
    "    \n",
    "    ## Make ensembl id and official symbol the first two columns \n",
    "    dff = dff[dff.columns[-2:].tolist() + dff.columns[:-2].tolist()].copy()\n",
    "\n",
    "    \n",
    "    return dff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a5325e4f-507b-40be-a6a0-7d4e88d7f212",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Import data\n",
    "\n",
    "df = pd.read_csv(\"../../data/raw_data/meta-analysis/marques-coelho_et_al.csv\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "aeadfa4f-62a4-44f2-aa8e-8942e10d1c9d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TXNAME</th>\n",
       "      <th>dataset</th>\n",
       "      <th>GENEID</th>\n",
       "      <th>GENE_NAME</th>\n",
       "      <th>gene.log2FC</th>\n",
       "      <th>gene.padj</th>\n",
       "      <th>transcript.log2FC</th>\n",
       "      <th>transcript.padj</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>iso_biotype</th>\n",
       "      <th>dtu.dIF</th>\n",
       "      <th>dtu.gene.padj</th>\n",
       "      <th>dtu.isoform.padj</th>\n",
       "      <th>dtu.ofdr.gene</th>\n",
       "      <th>dtu.ofdr.transcript</th>\n",
       "      <th>threshDT</th>\n",
       "      <th>threshG</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000000233.9</td>\n",
       "      <td>MAYO</td>\n",
       "      <td>ENSG00000004059.10</td>\n",
       "      <td>ARF5</td>\n",
       "      <td>-0.281214</td>\n",
       "      <td>0.000813</td>\n",
       "      <td>-0.283385</td>\n",
       "      <td>0.011291</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>0.004370</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not Significant</td>\n",
       "      <td>Non-Significant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000000233.9</td>\n",
       "      <td>MSBB BM10</td>\n",
       "      <td>ENSG00000004059.10</td>\n",
       "      <td>ARF5</td>\n",
       "      <td>-0.263954</td>\n",
       "      <td>0.031766</td>\n",
       "      <td>-0.243700</td>\n",
       "      <td>0.126583</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>0.004464</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not Significant</td>\n",
       "      <td>Non-Significant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000000233.9</td>\n",
       "      <td>MSBB BM22</td>\n",
       "      <td>ENSG00000004059.10</td>\n",
       "      <td>ARF5</td>\n",
       "      <td>-0.260638</td>\n",
       "      <td>0.050646</td>\n",
       "      <td>-0.291972</td>\n",
       "      <td>0.096546</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>-0.001508</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not Significant</td>\n",
       "      <td>Non-Significant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000000233.9</td>\n",
       "      <td>MSBB BM36</td>\n",
       "      <td>ENSG00000004059.10</td>\n",
       "      <td>ARF5</td>\n",
       "      <td>-0.435165</td>\n",
       "      <td>0.000253</td>\n",
       "      <td>-0.472171</td>\n",
       "      <td>0.007999</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>-0.053296</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DEG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000000233.9</td>\n",
       "      <td>MSBB BM44</td>\n",
       "      <td>ENSG00000004059.10</td>\n",
       "      <td>ARF5</td>\n",
       "      <td>-0.252960</td>\n",
       "      <td>0.041865</td>\n",
       "      <td>-0.304051</td>\n",
       "      <td>0.075831</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>-0.003478</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not Significant</td>\n",
       "      <td>Non-Significant</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              TXNAME    dataset              GENEID GENE_NAME  gene.log2FC  \\\n",
       "0  ENST00000000233.9       MAYO  ENSG00000004059.10      ARF5    -0.281214   \n",
       "1  ENST00000000233.9  MSBB BM10  ENSG00000004059.10      ARF5    -0.263954   \n",
       "2  ENST00000000233.9  MSBB BM22  ENSG00000004059.10      ARF5    -0.260638   \n",
       "3  ENST00000000233.9  MSBB BM36  ENSG00000004059.10      ARF5    -0.435165   \n",
       "4  ENST00000000233.9  MSBB BM44  ENSG00000004059.10      ARF5    -0.252960   \n",
       "\n",
       "   gene.padj  transcript.log2FC  transcript.padj    gene_biotype  \\\n",
       "0   0.000813          -0.283385         0.011291  protein_coding   \n",
       "1   0.031766          -0.243700         0.126583  protein_coding   \n",
       "2   0.050646          -0.291972         0.096546  protein_coding   \n",
       "3   0.000253          -0.472171         0.007999  protein_coding   \n",
       "4   0.041865          -0.304051         0.075831  protein_coding   \n",
       "\n",
       "      iso_biotype   dtu.dIF  dtu.gene.padj  dtu.isoform.padj  dtu.ofdr.gene  \\\n",
       "0  protein_coding  0.004370            NaN               NaN            NaN   \n",
       "1  protein_coding  0.004464            NaN               NaN            NaN   \n",
       "2  protein_coding -0.001508            NaN               NaN            NaN   \n",
       "3  protein_coding -0.053296            NaN               NaN            NaN   \n",
       "4  protein_coding -0.003478            NaN               NaN            NaN   \n",
       "\n",
       "   dtu.ofdr.transcript         threshDT          threshG  \n",
       "0                  NaN  Not Significant  Non-Significant  \n",
       "1                  NaN  Not Significant  Non-Significant  \n",
       "2                  NaN  Not Significant  Non-Significant  \n",
       "3                  NaN              NaN              DEG  \n",
       "4                  NaN  Not Significant  Non-Significant  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
